independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -68,17 +69,22 @@ public void init(ClassLoader runnerClassLoader) {
     }
 
     @Override
-    public byte[] getResourceData(String resource) {
+    public ByteBuffer getResourceData(String resource, ByteBuffer pooledHeapBuffer) {
         final ZipFile zipFile = readLockAcquireAndGetJarReference();
         try {
             ZipEntry entry = zipFile.getEntry(resource);
             if (entry == null) {
                 return null;
             }
             try (InputStream is = zipFile.getInputStream(entry)) {
-                byte[] data = new byte[(int) entry.getSize()];
+                int rem = (int) entry.getSize();
+                if (pooledHeapBuffer == null || pooledHeapBuffer.capacity() < rem) {
+                    pooledHeapBuffer = ByteBuffer.allocate(rem);
```

Comment:
Maybe in this case it could be better - or at least somewhat less code - to do something like `pooledHeapBuffer = ByteBuffer.wrap(is.readAllBytes());`, and only do the unrolled read down below if the buffer was recycled?

Reply:
Funny thing, `ZipInputStream` unconditionally has an internal 512-byte buffer. Also they don't override `readAllBytes` 😬

Reply:
Maybe we need our own version 😎

Reply:
For Jandex with @Ladicek we've used few tricks to avoid the internal allocations of similar classes (I have to check which ones....) - and reuse them over and over.
It wasn't the part which improved the throughput BUT it reduces the allocation pressure; which is what determine the initial heap size (and RSS) unless the user pre-size it.